File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/MusicPlayer.py
import random

class MusicPlayer:
    def __init__(self):
        self.playlist = []
        self.currentSong = None
        self.volume = 50

    def add_song(self, song):
        self.playlist.append(song)

    def remove_song(self, song):
        if song in self.playlist:
            self.playlist.remove(song)
            if self.currentSong == song:
                self.stop()

    def play(self):
        if not self.playlist:
            return None
        if self.currentSong is not None:
            for song in self.playlist:
                if song == self.currentSong:
                    return self.currentSong
        return self.playlist[0] if self.playlist else None

    def stop(self):
        if self.currentSong is not None:
            self.currentSong = None
            return True
        return False

    def switch_song(self):
        if self.currentSong is None:
            return False
        try:
            index = self.playlist.index(self.currentSong)
        except ValueError:
            return False
        if index < len(self.playlist) - 1:
            self.currentSong = self.playlist[index + 1]
            return True
        return False

    def previous_song(self):
        if self.currentSong is None:
            return False
        try:
            index = self.playlist.index(self.currentSong)
        except ValueError:
            return False
        if index > 0:
            self.currentSong = self.playlist[index - 1]
            return True
        return False

    def set_volume(self, volume):
        if 0 <= volume <= 100:
            self.volume = volume
            return True
        return False

    def shuffle(self):
        if not self.playlist:
            return False
        random.shuffle(self.playlist)
        return True

    def get_playlist(self):
        return self.playlist

    def set_playlist(self, playlist):
        self.playlist = playlist

    def get_current_song(self):
        return self.currentSong

    def set_current_song(self, currentSong):
        self.currentSong = currentSong

    def get_volume(self):
        return self.volume

File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/MusicPlayer.cpp
#include <vector>
#include <string>
#include <algorithm>
#include <random>
#include <optional>

class MusicPlayer {
private:
    std::vector<std::string> playlist;
    std::optional<std::string> currentSong;
    int volume;

public:
    MusicPlayer()
        : playlist(), currentSong(std::nullopt), volume(50) {}

    void addSong(const std::string& song) {
        playlist.push_back(song);
    }

    void removeSong(const std::string& song) {
        auto it = std::find(playlist.begin(), playlist.end(), song);
        if (it != playlist.end()) {
            playlist.erase(it);
            if (currentSong.has_value() && currentSong.value() == song) {
                stop();
            }
        }
    }

    std::optional<std::string> play() {
        if (playlist.empty()) {
            return std::nullopt;
        }
        if (currentSong.has_value()) {
            if (std::find(playlist.begin(), playlist.end(), currentSong.value()) != playlist.end()) {
                return currentSong;
            }
        }
        return playlist[0];
    }

    bool stop() {
        if (currentSong.has_value()) {
            currentSong = std::nullopt;
            return true;
        } else {
            return false;
        }
    }

    bool switchSong() {
        if (!currentSong.has_value()) {
            return false;
        }
        auto it = std::find(playlist.begin(), playlist.end(), currentSong.value());
        if (it == playlist.end()) {
            return false; // should not happen normally
        }
        size_t index = std::distance(playlist.begin(), it);
        if (index < playlist.size() - 1) {
            currentSong = playlist[index + 1];
            return true;
        } else {
            return false;
        }
    }

    bool previousSong() {
        if (!currentSong.has_value()) {
            return false;
        }
        auto it = std::find(playlist.begin(), playlist.end(), currentSong.value());
        if (it == playlist.end()) {
            return false;
        }
        size_t index = std::distance(playlist.begin(), it);
        if (index > 0) {
            currentSong = playlist[index - 1];
            return true;
        } else {
            return false;
        }
    }

    bool setVolume(int vol) {
        if (vol >= 0 && vol <= 100) {
            volume = vol;
            return true;
        } else {
            return false;
        }
    }

    bool shuffle() {
        if (playlist.empty()) {
            return false;
        }
        std::random_device rd;
        std::mt19937 g(rd());
        std::shuffle(playlist.begin(), playlist.end(), g);
        return true;
    }

    std::vector<std::string>& getPlaylist() {
        return playlist;
    }

    void setPlaylist(const std::vector<std::string>& newPlaylist) {
        playlist = newPlaylist;
    }

    std::optional<std::string> getCurrentSong() const {
        return currentSong;
    }

    void setCurrentSong(const std::optional<std::string>& song) {
        currentSong = song;
    }

    int getVolume() const {
        return volume;
    }
};